content/blog/2019-12-09-comp-cert/index.md

```diff
@@ -0,0 +1,111 @@
+
+## Evaluation
+
+In general CompCert is able to output code with similar performance characteristics to gcc at -O1 and -O2.
+
+<img src="perf.png" style="width: 100%">
+
+However, we think a more important metric is the correctness of CompCert, since this was the primary purpose of creating it. This is something the paper was not able to do because nobody had seriously tested CompCert at the time of the paper's release. However, several automated compiler testing tools such as [Csmith](https://www.cs.utah.edu/~regehr/papers/pldi11-preprint.pdf) and Orion (from the [Equivalence Module Inputs](http://vuminhle.com/pdf/pldi14-emi.pdf) paper) have reported a handful of bugs in CompCert over the years. We looked into issues on the official CompCert github page and bug reports generated by Orion to try and figure out what some of the bugs were and where the manifested themselves.
+
+The github repository for [CompCert](https://github.com/AbsInt/CompCert) seems to have been created in 2015, while the paper is dated 2009. So there may be several bugs found in the original version of CompCert that were patched when the github repository was started. The Orion project includes a [page](https://web.cs.ucdavis.edu/~su/emi-project/compcert.html) with all of the bugs it found that led to open issues on the CompCert github repo. There are 31 total issues on this list, with 27 of them being marked as fixed and the remaining 4 marked as won't fix. This list of issues was reported between August 2016 and May 2017. Most of the issues reported seem to be issues in the front end OCaml code and are mostly crash failures. There are a few crash failures associated with the unverified register allocation code that seem to have required some updates to the Coq proofs such as [issue 183](https://github.com/AbsInt/CompCert/issues/183). However, it does not appear as if there was a case where CompCert silently generated a miscompilation (one case was mentioned in the Csmith paper but the culprit was determined to be the front end parser). This suggests that CompCert does indeed succeed at its goal of creating a C compiler with no miscompilation bugs.
```

Comment:
Indeed, as the Csmith paper says, they found bugs in CompCert, *but they were in the unverified components* such as the parser. (The paper also remarks that Csmith's bug reports motivated the CompCert people to verify more of the stuff (aka "shrink the TCB" for CompCert).) As far as I am aware, no one has ever found a bug in the "verified parts" of CompCert, which could only really indicate that (a) the specs were somehow wrong, or (b) there is a bug in Coq itself, both of which are pretty implausible.

Also, my reading of the Csmith paper is that that bug wasn't in the parser exactly. The paper says:

> This bug and five others like it were in CompCert’s unverified front-end code. Partly in response to these bug reports, the main CompCert developer expanded the verified portion of CompCert to include C’s integer promotions and other tricky implicit casts.

It never says it was in the parser, and the implication from what was then subsequently verified implies that it might have been in a few AST simplifications they did before the verified part began.